core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala

```diff
@@ -315,130 +316,148 @@ trait SubmissionMonitor extends FutureSupport with LazyLogging with RawlsInstrum
-                }
-                handleOutputs(batch, dataAccess, innerSpan)
-              }
-            }))
-          }
-          .unsafeToFuture()
-          .recoverWith {
-            // If there is something fatally wrong handling outputs for any workflow, mark all the workflows as failed
-            case fatal: RawlsFatalExceptionWithErrorReport =>
-              markWorkflowsFailed(allWorkflows, fatal)
-          }
-      } flatMap { _ =>
-        // NEW TXN! Update statuses for workflows and submission.
-        datasource.inTransaction { dataAccess =>
-          // Refetch workflows as some may have been marked as Failed by handleOutputs.
-          dataAccess.workflowQuery.findWorkflowByIds(allWorkflows.map(_.id)).result flatMap { updatedRecs =>
-            // New statuses according to the execution service.
-            val workflowIdToNewStatus = allWorkflows.map(workflowRec => workflowRec.id -> workflowRec.status).toMap
+      logger.info(
+        s"will process ${noOutputs.size} workflow(s) without outputs ($noOutputsGrouped) and ${yesOutputs.size} workflow(s) with outputs ($yesOutputsGrouped) for submission $submissionId"
+      )
 
-            // No need to update statuses for any workflows that are in terminal statuses.
-            // Doing so would potentially overwrite them with the execution service status if they'd been marked as failed by attachOutputs.
-            val workflowsToUpdate = updatedRecs.filter(rec =>
-              !WorkflowStatuses.terminalStatuses.contains(WorkflowStatuses.withName(rec.status))
-            )
-            val workflowsWithNewStatuses =
-              workflowsToUpdate.map(rec => rec.copy(status = workflowIdToNewStatus(rec.id)))
-
-            // to minimize database updates batch 1 update per workflow status
-            DBIO.sequence(workflowsWithNewStatuses.groupBy(_.status).map { case (status, recs) =>
-              dataAccess.workflowQuery.batchUpdateStatus(recs, WorkflowStatuses.withName(status))
-            })
-
-          } flatMap { _ =>
-            // update submission after workflows are updated
-            updateSubmissionStatus(dataAccess) map { shouldStop: Boolean =>
-              // return a message about whether our submission is done entirely
-              StatusCheckComplete(shouldStop)
+      /**
+        * Reusable function for looping over a group of workflows, using IO to ensure operations happen sequentially,
+        * and calling processWorkflow() for each workflow in the group
+        * @param workflowGroup the workflows to process
+        * @param tracingLabel human-readable label for tracing purposes
+        * @return nothing interesting
+        */
+      def processMultipleWorkflows(
+        workflowGroup: Seq[(WorkflowRecord, Option[ExecutionServiceOutputs])],
+        tracingLabel: String
+      ) =
+        traceWithParent(tracingLabel, rootSpan) { innerSpan =>
```

Comment:
I think this trace applies to the multiple workflows. Is it an anti-pattern to spin up too many sub-traces (eg one per workflow)?

Reply:
at the `handleOutputs` level, we are adding a trace per workflow. I fear it might be too much - if those traces are unwieldy in practice we'll have to tweak after the fact.